File: source/test_stand/motor_position.py
#!/usr/bin/env python3
"""
Test Motor Control with Buttons
Uses red and green buttons to control DC motor direction via DRV8833.
"""

import time
import logging
import os
from button_driver import Buttons
from motor_driver import MotorController


def main():
    """
    Main function to control motor with buttons.
    Red button: motor runs forward
    Green button: motor runs reverse
    No button: motor stops
    """
    # Setup logging
    log_level = os.getenv("LOG_LEVEL", "INFO").upper()
    logging.basicConfig(
        level=getattr(logging, log_level, logging.INFO),
        format='%(asctime)s - %(levelname)s - %(message)s'
    )

    logging.info("Starting button-controlled motor test")

    # Initialize components
    buttons = Buttons(pull_up=True)
    # mode_already_set=True because Buttons class already called GPIO.setmode()
    # motor = MotorController(ain1_gpio=4, ain2_gpio=5, mode_already_set=True)
    # motor = MotorController(ain1_gpio=6, ain2_gpio=12, mode_already_set=True)
    motor = MotorController(ain1_gpio=13, ain2_gpio=16, mode_already_set=True)

    try:
        logging.info("Press RED button for forward, GREEN button for reverse")
        logging.info("Press Ctrl+C to exit")
        last_state = "stopped"
        while True:
            red_pressed = buttons.is_red_button_pressed()
            green_pressed = buttons.is_green_button_pressed()

            if red_pressed and not green_pressed:
                if last_state != "forward":
                    motor.forward(speed=15)
                    logging.info("Motor running FORWARD at 15% speed (red button pressed)")
                    last_state = "forward"

            elif green_pressed and not red_pressed:
                if last_state != "reverse":
                    motor.reverse(speed=15)
                    logging.info("Motor running REVERSE at 15% speed (green button pressed)")
                    last_state = "reverse"

            elif red_pressed and green_pressed:
                # Both buttons pressed - stop for safety
                if last_state != "stopped":
                    motor.stop()
                    logging.warning("Both buttons pressed - motor STOPPED")
                    last_state = "stopped"

            else:
                # No buttons pressed
                if last_state != "stopped":
                    motor.stop()
                    logging.info("No buttons pressed - motor STOPPED")
                    last_state = "stopped"

            # Small delay to prevent excessive CPU usage
            time.sleep(0.05)

    except KeyboardInterrupt:
        logging.info("Program interrupted by user")

    finally:
        # Clean up
        motor.stop()
        buttons.cleanup()
        logging.info("Cleanup complete")


if __name__ == "__main__":
    main()

File: source/test_stand_c/motor_position.cpp
/**
 * @file motor_position.cpp
 * @brief Test Motor Control with Joystick
 *
 * Uses joystick to control three DC motors for test arm joints via DRV8833.
 */

#include "JoystickController.h"
#include "MotorController.h"
#include "ButtonDriver.h"
#include <iostream>
#include <unistd.h>
#include <csignal>
#include <cmath>

volatile sig_atomic_t stop_flag = 0;

void signal_handler(int signum) {
    stop_flag = 1;
}

int main() {
    // Setup signal handler for Ctrl+C
    signal(SIGINT, signal_handler);

    std::cout << "Starting joystick-controlled motor test" << std::endl;

    try {
        // Initialize joystick
        JoystickController joystick(0, 1, 17);

        // Calibrate the joystick
        std::cout << "Calibrating joystick..." << std::endl;
        joystick.calibrate();

        // Initialize buttons
        // mode_already_set=true because JoystickController already initialized pigpio
        Buttons buttons(true);

        // Initialize three motors
        // mode_already_set=true because JoystickController already initialized pigpio
        MotorController motor_knee_x(13, 16, true);
        MotorController motor_knee_y(6, 12, true);
        MotorController motor_wrist(4, 5, true);

        std::cout << "Motors initialized:" << std::endl;
        std::cout << "  motor_knee_x:  GPIO 13/16" << std::endl;
        std::cout << "  motor_knee_y:  GPIO 6/12" << std::endl;
        std::cout << "  motor_wrist:   GPIO 4/5" << std::endl;

        std::cout << "Joystick X controls motor_knee_x" << std::endl;
        std::cout << "Joystick Y controls motor_knee_y" << std::endl;
        std::cout << "Green button controls motor_wrist forward at 20%" << std::endl;
        std::cout << "Red button controls motor_wrist reverse at 20%" << std::endl;
        std::cout << "Press Ctrl+C to exit" << std::endl;

        int last_x_speed = 0;
        int last_y_speed = 0;
        bool last_green_pressed = false;
        bool last_red_pressed = false;

        while (!stop_flag) {
            // Read joystick position
            double x_pos = joystick.read_x();  // -1.0 to 1.0
            double y_pos = joystick.read_y();  // -1.0 to 1.0

            // Map joystick position to motor speed
            // At neutral (0.0), motor stops
            // At full deflection (±1.0), motor runs at specified speed
            int x_speed = static_cast<int>(x_pos * 20);  // Maps -1.0 to 1.0 => -20 to 20
            int y_speed = static_cast<int>(y_pos * 40);  // Maps -1.0 to 1.0 => -40 to 40

            // Control motor_knee_x based on joystick X (reversed direction)
            if (x_speed != last_x_speed) {
                if (x_speed > 0) {
                    motor_knee_x.reverse(std::abs(x_speed));
                    std::cout << "motor_knee_x: REVERSE at " << std::abs(x_speed)
                              << "% (X=" << std::fixed << std::setprecision(2) << x_pos << ")" << std::endl;
                } else if (x_speed < 0) {
                    motor_knee_x.forward(std::abs(x_speed));
                    std::cout << "motor_knee_x: FORWARD at " << std::abs(x_speed)
                              << "% (X=" << std::fixed << std::setprecision(2) << x_pos << ")" << std::endl;
                } else {
                    motor_knee_x.stop();
                    std::cout << "motor_knee_x: STOP (X=0.00)" << std::endl;
                }
                last_x_speed = x_speed;
            }

            // Control motor_knee_y based on joystick Y
            if (y_speed != last_y_speed) {
                if (y_speed > 0) {
                    motor_knee_y.forward(std::abs(y_speed));
                    std::cout << "motor_knee_y: FORWARD at " << std::abs(y_speed)
                              << "% (Y=" << std::fixed << std::setprecision(2) << y_pos << ")" << std::endl;
                } else if (y_speed < 0) {
                    motor_knee_y.reverse(std::abs(y_speed));
                    std::cout << "motor_knee_y: REVERSE at " << std::abs(y_speed)
                              << "% (Y=" << std::fixed << std::setprecision(2) << y_pos << ")" << std::endl;
                } else {
                    motor_knee_y.stop();
                    std::cout << "motor_knee_y: STOP (Y=0.00)" << std::endl;
                }
                last_y_speed = y_speed;
            }

            // Read button states
            bool green_pressed = buttons.is_green_button_pressed();
            bool red_pressed = buttons.is_red_button_pressed();

            // Control motor_wrist based on buttons
            if (green_pressed != last_green_pressed) {
                if (green_pressed) {
                    motor_wrist.forward(20);
                    std::cout << "motor_wrist: FORWARD at 20% (Green button pressed)" << std::endl;
                } else {
                    motor_wrist.stop();
                    std::cout << "motor_wrist: STOP (Green button released)" << std::endl;
                }
                last_green_pressed = green_pressed;
            }

            if (red_pressed != last_red_pressed) {
                if (red_pressed) {
                    motor_wrist.reverse(20);
                    std::cout << "motor_wrist: REVERSE at 20% (Red button pressed)" << std::endl;
                } else {
                    motor_wrist.stop();
                    std::cout << "motor_wrist: STOP (Red button released)" << std::endl;
                }
                last_red_pressed = red_pressed;
            }

            // Small delay to prevent excessive CPU usage
            usleep(50000);  // 50ms
        }

        std::cout << "\nProgram interrupted by user" << std::endl;

        // Clean up - stop all motors
        std::cout << "Stopping all motors..." << std::endl;
        motor_knee_x.stop();
        motor_knee_y.stop();
        motor_wrist.stop();

        std::cout << "Cleanup complete" << std::endl;

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        gpioTerminate();
        return 1;
    }

    // Cleanup pigpio
    gpioTerminate();

    return 0;
}
